File: src/vellaris/core/wrap.py
"""Passphrase-wrapped private-key blobs.

Combines :mod:`vellaris.core.kdf` and :mod:`vellaris.core.symmetric` to
turn a PEM-encoded private key into a single bytestring that can be
written to ``~/.vellaris/keys/<user-id>.key`` or pushed (opt-in) as an
opaque blob to the server. The server cannot decrypt the result because
the passphrase never leaves the client.

Layout for ``WRAPPED_V1``::

    ┌─────────┬──────────┬──────────────┬──────────────┬──────────────────┐
    │ version │   salt   │ params_len   │ params_json  │  inner ciphertext│
    │  1 byte │ 16 bytes │   2 bytes    │   variable   │   (wire envelope)│
    └─────────┴──────────┴──────────────┴──────────────┴──────────────────┘

The version byte, salt, and params_json bytes are bound to the AES-GCM
ciphertext via the AEAD's associated-data slot, so altering them after
wrapping invalidates the tag and produces a :class:`DecryptError` at
unwrap time.
"""

from __future__ import annotations

import json
import struct
from typing import Final

from vellaris.core.errors import DecryptError, WireFormatError
from vellaris.core.kdf import SALT_SIZE, Argon2Params, derive_key, random_salt
from vellaris.core.symmetric import decrypt as aes_decrypt
from vellaris.core.symmetric import encrypt
from vellaris.core.wire import pack, unpack

WRAPPED_V1: Final[int] = 0x01

_PARAMS_LEN_FIELD: Final[int] = 2  # uint16 big-endian
_HEADER_FIXED: Final[int] = 1 + SALT_SIZE + _PARAMS_LEN_FIELD


def _associated_data(version: int, salt: bytes, params_bytes: bytes) -> bytes:
    """Bind version + salt + params to the AEAD tag."""
    return bytes([version]) + salt + params_bytes


def wrap_private_key(
    pem_bytes: bytes,
    passphrase: bytes | str,
    *,
    params: Argon2Params | None = None,
) -> bytes:
    """Encrypt ``pem_bytes`` with a key derived from ``passphrase``.

    Returns a single bytestring suitable for writing to disk. The salt
    and Argon2 parameters are stored alongside so the same passphrase
    can re-derive the key at unwrap time.
    """
    if not isinstance(pem_bytes, (bytes, bytearray)):
        raise TypeError(f"pem_bytes must be bytes, got {type(pem_bytes).__name__}")

    p = params or Argon2Params()
    salt = random_salt()
    key = derive_key(passphrase, salt, p)
    params_bytes = json.dumps(p.to_dict(), sort_keys=True, separators=(",", ":")).encode("utf-8")
    if len(params_bytes) > 0xFFFF:
        # Practically unreachable; defensive cap to keep the 2-byte length field honest.
        raise WireFormatError(f"params_json too large: {len(params_bytes)} bytes")

    aad = _associated_data(WRAPPED_V1, salt, params_bytes)
    sealed = encrypt(bytes(pem_bytes), key, associated_data=aad)
    inner = pack(sealed)

    return bytes([WRAPPED_V1]) + salt + struct.pack(">H", len(params_bytes)) + params_bytes + inner


def unwrap_private_key(blob: bytes, passphrase: bytes | str) -> bytes:
    """Decrypt a blob produced by :func:`wrap_private_key`.

    Raises :class:`WireFormatError` for structural problems (unknown
    version, truncated header, malformed params) and :class:`DecryptError`
    for crypto failures (wrong passphrase, tampered bytes).
    """
    if not isinstance(blob, (bytes, bytearray)):
        raise WireFormatError(f"blob must be bytes, got {type(blob).__name__}")
    if len(blob) < _HEADER_FIXED:
        raise WireFormatError(f"blob too short: {len(blob)} < {_HEADER_FIXED}")

    version = blob[0]
    if version != WRAPPED_V1:
        raise WireFormatError(f"unknown wrapped-key version: 0x{version:02x}")

    salt = bytes(blob[1 : 1 + SALT_SIZE])
    (params_len,) = struct.unpack(">H", bytes(blob[1 + SALT_SIZE : _HEADER_FIXED]))

    params_end = _HEADER_FIXED + params_len
    if params_end > len(blob):
        raise WireFormatError(f"truncated params field: needed {params_end} bytes, got {len(blob)}")

    params_bytes = bytes(blob[_HEADER_FIXED:params_end])
    inner = bytes(blob[params_end:])

    try:
        params_dict = json.loads(params_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise WireFormatError(f"malformed params JSON: {exc}") from exc

    params = Argon2Params.from_dict(params_dict)
    sealed = unpack(inner)

    key = derive_key(passphrase, salt, params)
    aad = _associated_data(version, salt, params_bytes)
    return aes_decrypt(sealed, key, associated_data=aad)


def is_wrapped_blob(blob: bytes) -> bool:
    """Cheap sniff: does ``blob`` start with a known wrapped-key version byte?"""
    return len(blob) >= 1 and blob[0] == WRAPPED_V1


__all__ = [
    "WRAPPED_V1",
    "DecryptError",  # re-exported for callers that want a one-stop import
    "is_wrapped_blob",
    "unwrap_private_key",
    "wrap_private_key",
]

File: web/src/crypto/wrap.ts
/**
 * Passphrase-wrapped private-key blobs.
 *
 * Mirrors `src/vellaris/core/wrap.py`. Combines {@link "./kdf".deriveKey}
 * with {@link "./symmetric".encrypt} to turn a PEM-encoded private key
 * into a single bytestring storable in localStorage / IndexedDB or
 * pushable (opt-in) to the server as an opaque blob. The server cannot
 * decrypt the result because the passphrase never leaves the client.
 *
 * Layout for `WRAPPED_V1`:
 *
 *   ┌─────────┬──────────┬──────────────┬──────────────┬──────────────────┐
 *   │ version │   salt   │ params_len   │ params_json  │  inner ciphertext│
 *   │  1 byte │ 16 bytes │   2 bytes    │   variable   │   (wire envelope)│
 *   └─────────┴──────────┴──────────────┴──────────────┴──────────────────┘
 *
 * `params_json` is the JSON encoding of {@link Argon2ParamsDict} with
 * `sort_keys=True, separators=(',', ':')` — Python's canonical compact form.
 * The version byte, salt, and params bytes are bound to the AES-GCM
 * ciphertext via the AEAD's associated-data slot, so altering them after
 * wrapping invalidates the tag and produces a `DecryptError` at unwrap time.
 */

import { WireFormatError } from './errors.ts'
import {
  type Argon2Params,
  DEFAULT_PARAMS,
  SALT_SIZE,
  deriveKey,
  paramsFromDict,
  paramsToDict,
  randomSalt,
} from './kdf.ts'
import { decrypt as aesDecrypt, encrypt as aesEncrypt } from './symmetric.ts'
import { pack, unpack } from './wire.ts'

export const WRAPPED_V1 = 0x01

const PARAMS_LEN_FIELD = 2 // uint16 big-endian
const HEADER_FIXED = 1 + SALT_SIZE + PARAMS_LEN_FIELD

/**
 * Canonical JSON encoding for Argon2 params — matches Python's
 * `json.dumps(p.to_dict(), sort_keys=True, separators=(',', ':'))`.
 *
 * The dict has fixed keys (`l, m, p, t` after sorting) so we hand-roll
 * the encoder instead of pulling in a canonical-json library.
 */
function encodeParamsJson(p: Argon2Params): Uint8Array {
  const dict = paramsToDict(p)
  // Sorted alphabetically: l, m, p, t — matches Python sort_keys=True.
  const json = `{"l":${dict.l},"m":${dict.m},"p":${dict.p},"t":${dict.t}}`
  return new TextEncoder().encode(json)
}

function associatedData(version: number, salt: Uint8Array, paramsBytes: Uint8Array): Uint8Array {
  const out = new Uint8Array(1 + salt.length + paramsBytes.length)
  out[0] = version
  out.set(salt, 1)
  out.set(paramsBytes, 1 + salt.length)
  return out
}

/**
 * Encrypt `pemBytes` with a key derived from `passphrase`.
 *
 * Returns a single bytestring suitable for writing to localStorage. The
 * salt and Argon2 parameters are stored alongside so the same passphrase
 * can re-derive the key at unwrap time.
 */
export async function wrapPrivateKey(
  pemBytes: Uint8Array,
  passphrase: Uint8Array | string,
  options?: { params?: Argon2Params; salt?: Uint8Array },
): Promise<Uint8Array> {
  const params = options?.params ?? DEFAULT_PARAMS
  const salt = options?.salt ?? randomSalt()
  if (salt.length !== SALT_SIZE) {
    throw new WireFormatError(`salt must be ${SALT_SIZE} bytes, got ${salt.length}`)
  }

  const key = await deriveKey(passphrase, salt, params)
  const paramsBytes = encodeParamsJson(params)
  if (paramsBytes.length > 0xffff) {
    throw new WireFormatError(`params_json too large: ${paramsBytes.length} bytes`)
  }

  const aad = associatedData(WRAPPED_V1, salt, paramsBytes)
  const sealed = await aesEncrypt(pemBytes, key, { associatedData: aad })
  const inner = pack(sealed)

  const out = new Uint8Array(HEADER_FIXED + paramsBytes.length + inner.length)
  let offset = 0
  out[offset++] = WRAPPED_V1
  out.set(salt, offset)
  offset += SALT_SIZE
  // params_len as uint16 big-endian
  out[offset++] = (paramsBytes.length >> 8) & 0xff
  out[offset++] = paramsBytes.length & 0xff
  out.set(paramsBytes, offset)
  offset += paramsBytes.length
  out.set(inner, offset)
  return out
}

/**
 * Decrypt a blob produced by {@link wrapPrivateKey}.
 *
 * Throws {@link WireFormatError} for structural problems and
 * {@link DecryptError} (re-thrown from {@link aesDecrypt}) for crypto
 * failures (wrong passphrase, tampered bytes).
 */
export async function unwrapPrivateKey(
  blob: Uint8Array,
  passphrase: Uint8Array | string,
): Promise<Uint8Array> {
  if (blob.length < HEADER_FIXED) {
    throw new WireFormatError(`blob too short: ${blob.length} < ${HEADER_FIXED}`)
  }
  const version = blob[0]
  if (version !== WRAPPED_V1) {
    throw new WireFormatError(
      `unknown wrapped-key version: 0x${version.toString(16).padStart(2, '0')}`,
    )
  }
  const salt = blob.slice(1, 1 + SALT_SIZE)
  const paramsLen = (blob[1 + SALT_SIZE] << 8) | blob[1 + SALT_SIZE + 1]
  const paramsEnd = HEADER_FIXED + paramsLen
  if (paramsEnd > blob.length) {
    throw new WireFormatError(
      `truncated params field: needed ${paramsEnd} bytes, got ${blob.length}`,
    )
  }
  const paramsBytes = blob.slice(HEADER_FIXED, paramsEnd)
  const inner = blob.slice(paramsEnd)

  let paramsDict: unknown
  try {
    paramsDict = JSON.parse(new TextDecoder().decode(paramsBytes))
  } catch (cause) {
    throw new WireFormatError(`malformed params JSON: ${(cause as Error).message}`, { cause })
  }
  if (typeof paramsDict !== 'object' || paramsDict === null) {
    throw new WireFormatError('params JSON must be an object')
  }
  const params = paramsFromDict(paramsDict as Record<string, unknown>)
  const sealed = unpack(inner)

  const key = await deriveKey(passphrase, salt, params)
  const aad = associatedData(version, salt, paramsBytes)
  return aesDecrypt(sealed, key, { associatedData: aad })
}

/** Cheap sniff: does `blob` start with a known wrapped-key version byte? */
export function isWrappedBlob(blob: Uint8Array): boolean {
  return blob.length >= 1 && blob[0] === WRAPPED_V1
}
